pkg/pool-hooks/contracts/utils/StableSurgeMedianMath.sol

```diff
@@ -0,0 +1,58 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+pragma solidity ^0.8.24;
+
+library StableSurgeMedianMath {
+    function calculateImbalance(uint256[] memory balances) internal pure returns (uint256) {
+        uint256[] memory sortedBalances = sort(balances);
+        uint256 median = findMedian(sortedBalances);
+
+        uint256 totalBalance = 0;
+        uint256 totalDiff = 0;
+
+        for (uint i = 0; i < balances.length; i++) {
+            totalBalance += balances[i];
+            totalDiff += absSub(balances[i], median);
+        }
+
+        return (totalDiff * 1e18) / totalBalance;
+    }
+
+    function findMedian(uint256[] memory sortedBalances) internal pure returns (uint256) {
+        uint256 mid = sortedBalances.length / 2;
+
+        if (sortedBalances.length % 2 == 0) {
+            return (sortedBalances[mid - 1] + sortedBalances[mid]) / 2;
+        } else {
+            return sortedBalances[mid];
+        }
+    }
+
+    // First implementation had quickselect algorithm, but it was removed in the final version
```

Comment:
Note also the bubble sort in InputHelpers.sortTokens. How does that compare?

Reply:
This insert sort uses less gas than bubble sort. The worth case with 8-length arrays uses 15K gas. The average case is 10K.